datastream-bigquery/src/main/java/com/linkedin/datastream/bigquery/BigqueryBatchCommitter.java

```diff
@@ -195,80 +205,78 @@ public void commit(List<InsertAllRequest.RowToInsert> batch,
-                            this.getClass().getSimpleName(),
-                            recordMetadata.get(i).getTopic(),
-                            "errorCount",
-                            1);
-                    ackCallbacks.get(i).onCompletion(recordMetadata.get(i), exception);
-                } else {
-                    Long key = Long.valueOf(i);
-                    if (response != null && response.hasErrors() && response.getInsertErrors().containsKey(key)) {
-                        LOG.warn("Failed to insert a row {} {}", i, response.getInsertErrors().get(key));
-                        DynamicMetricsManager.getInstance().createOrUpdateMeter(
-                                this.getClass().getSimpleName(),
-                                recordMetadata.get(i).getTopic(),
-                                "errorCount",
-                                1);
-                        ackCallbacks.get(i).onCompletion(recordMetadata.get(i),
-                                new DatastreamTransientException(response.getInsertErrors().get(key).toString()));
-                    } else {
-                        DynamicMetricsManager.getInstance().createOrUpdateMeter(
-                                this.getClass().getSimpleName(),
-                                recordMetadata.get(i).getTopic(),
-                                "commitCount",
-                                1);
-                        DynamicMetricsManager.getInstance().createOrUpdateHistogram(
-                                this.getClass().getSimpleName(),
-                                recordMetadata.get(i).getTopic(),
-                                "eteLatency",
-                                System.currentTimeMillis() - sourceTimestamps.get(i));
-                        ackCallbacks.get(i).onCompletion(recordMetadata.get(i), null);
+            final String partition = partitionDateFormatter.format(LocalDate.now(ZoneOffset.UTC));
+            final String[] datasetTable = destination.split("/");
+            final String tableName = sanitizeTableName(datasetTable[1]);
+            final TableId tableId = TableId.of(datasetTable[0], String.format("%s$%s", tableName, partition));
+
+            LOG.debug("Committing a batch to dataset {} and table {}", datasetTable[0], tableName);
+
+            final InsertAllRequest insertAllRequest = InsertAllRequest.newBuilder(tableId, batch).build();
+
+            final long start = System.currentTimeMillis();
+            Map<Integer, Exception> insertErrors = insertRowsAndMapErrors(insertAllRequest);
+            final long end = System.currentTimeMillis();
+            DynamicMetricsManager.getInstance().createOrUpdateHistogram(classSimpleName, recordMetadata.get(0).getTopic(),
+                    "insertAllExecTime", end - start);
+
+            // If we encountered insert errors, try creating/updating the destination table before retrying
+            if (!insertErrors.isEmpty()) {
```

Comment:
It might be possible that the reason of an insert error is not caused by a schema mismatch? If so, we don't need to call `createOrUpdateTable(destination)`.
I would add a check to guarantee that the insertError is caused by a schema mismatch or a missing table

Reply:
I agree, it would be ideal if we could identify when a table didn't exist or had a schema mismatch before attempting the schema update. In my experience with Bigquery, it was not clear from the exceptions returned that the issue was caused by a schema mismatch. It seemed safest to perform the table schema check when encountering any exception until we have a better understanding of the exceptions we are encountering in production. The table schema check is a fairly low cost operation, since we're just getting table metadata and comparing the schemas before executing a more costly operation like creating/updating the table.